File: sem3/map/ToyLanguageInterpreter/src/utils/SymbolTableImpl.java
package utils;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by mirko on 12/10/2016.
 */
public class SymbolTableImpl<K, V> implements SymbolTable<K, V> {
    private Map<K, V> symbolTable;

    public SymbolTableImpl() {
        this.symbolTable = new HashMap<K, V>();
    }

    @Override
    public void add(K key, V value) {
        symbolTable.put(key, value);
    }

    @Override
    public boolean contains(K key) {
        return symbolTable.containsKey(key);
    }

    @Override
    public V getValue(K key) {
        if (!contains(key)) throw new RuntimeException("error: key \"" + key + "\" doesn't exist in symbol table");
        return symbolTable.get(key);
    }

    @Override
    public String toString() {
        StringBuilder string = new StringBuilder();
        string.append("SymbolTableImpl = {");

        if (!symbolTable.isEmpty()) string.append("\n");

        for (K key : symbolTable.keySet()) {
            string.append("   " + key + " <- " + symbolTable.get(key) + "\n");
        }

        string.append("}");
        return string.toString();
    }

    @Override
    public Iterable<Map.Entry<K, V>> getAll() {
        return symbolTable.entrySet();
    }
}

File: sem3/map/ToySharpLanguageInterpreter/ToySharpLanguageInterpreter/src/utils/SymbolTableImpl.cs
using System.Collections.Generic;
using System.Linq;

namespace ToySharpLanguageInterpreter
{
	public class SymbolTableImpl<K, V> : SymbolTable<K, V>
	{
		private Dictionary<K, V> symbolTable;

		public SymbolTableImpl()
		{
			this.symbolTable = new Dictionary<K, V>();
		}


		public void add(K key, V value)
		{
			if (symbolTable.ContainsKey(key)) symbolTable[key] = value;
			else symbolTable.Add(key, value);
		}


		public bool contains(K key)
		{
			return symbolTable.ContainsKey(key);
		}


		public V getValue(K key)
		{
			if (!contains(key)) throw (new InterpreterException("error: key \"" + key + "\" doesn't exist in symbol table"));
			return symbolTable[key];
		}


		public override string ToString()
		{
			string str = "";
			str += "SymbolTableImpl = {";

			if (symbolTable.Count() > 0) str += "\n";

			foreach (KeyValuePair<K, V> pair in symbolTable.ToList())
			{
				str += "   " + pair.Key + " <- " + pair.Value + "\n";
			}

			str += "}";
			return str;
		}


		public List<KeyValuePair<K, V>> getAll()
		{
			return symbolTable.ToList();
		}
	}
}